elements/lisk-chain/src/diff.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright © 2020 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+type HistoryType = [string?, number?];
+
+interface Frontier {
+	[key: string]: {
+		x: number;
+		history: HistoryType[];
+	};
+}
+
+const _diff = (initialBuffer: Buffer, finalBuffer: Buffer): HistoryType[] => {
+	const one = (idx: number): number => idx - 1;
+	const initial = Array.from(initialBuffer);
+	const final = Array.from(finalBuffer);
+	const initialBytesLength = initial.length;
+	const finalBytesLength = final.length;
+
+	if (initialBytesLength === 0) {
+		return final.map(b => ['+', b]);
+	}
+
+	if (finalBytesLength === 0) {
+		return initial.map(b => ['-', b]);
+	}
+
+	const frontier: Frontier = { 1: { x: 0, history: [] } };
+	for (let d = 0; d < initialBytesLength + finalBytesLength + 1; d += 1) {
+		for (let k = -d; k < d + 1; k += 2) {
+			let history: HistoryType[] = [];
+			const goDown =
+				k === -d || (k !== d && frontier[k - 1].x < frontier[k + 1].x);
+
+			let x: number;
+			if (goDown) {
+				history = frontier[k + 1].history;
+				x = frontier[k + 1].x;
+			} else {
+				history = frontier[k - 1].history;
+				x = frontier[k - 1].x + 1;
+			}
+
+			history = [...history];
```

Comment:
what is this doing?

Reply:
we want to preserve the old history because it might be used by other steps below

Reply:
but u are assigning to the same variable, so anyway it's overwritten...?😕

Reply:
So the history is a an array of arrays and that is stored in `frontier` object, so I think when we are looping through we don't want to change `frontier[k].history`, if we directly use `history = frontier[k].history`, and modify history then it will also change arrays in `frontier[k].history`.
To simply we could also do,

```typescript
if (goDown) {
				history = [...frontier[k + 1].history];
				x = frontier[k + 1].x;
			} else {
				history = [...frontier[k - 1].history];
				x = frontier[k - 1].x + 1;
			}
```

Reply:
make sense! yea i think it would be more clear if you do it on spot